llvm-spirv/lib/SPIRV/SPIRVRegularizeLLVM.cpp

```diff
@@ -288,6 +288,47 @@ void SPIRVRegularizeLLVMBase::expandVIDWithSYCLTypeByValComp(Function *F) {
       nullptr, &Attrs, true);
 }
 
+// intel/llvm customization
+void SPIRVRegularizeLLVMBase::finishSROACooperativeMatrix(Module *M) {
+  for (auto &F : *M) {
+    if (!F.getName().starts_with("_Z19__spirv_AccessChain"))
+      continue;
+    for (auto I = F.user_begin(), E = F.user_end(); I != E;) {
+      if (auto *CI = dyn_cast<CallInst>(*I++)) {
+        Instruction *Ptr =
+            dyn_cast<Instruction>(CI->getArgOperand(0)->stripPointerCasts());
+        StructType *WrapperMatrixTy =
+            dyn_cast<StructType>(cast<AllocaInst>(Ptr)->getAllocatedType());
```

Comment:
I guess that "CoopMatrix wrapped into a struct" covers all practical use cases, but I just wonder if it's possible to have any other nested type composed of CoopMatrix where this line would lead to a crash, because Argument #0 was not an alloca instruction wrt. the wrapped matrix type? In other words, would it make any sense to check `Ptr != nullptr` and `dyn_cast<AllocaInst>(Ptr)  != nullptr`?

Reply:
> because Argument #0 was not an alloca instruction wrt. the wrapped matrix type

Technically, it's unexpected, but yet possible if we store pointer in a GV and then load from it. I've added a sanity check for this. In real life if this appear translator will still fail to translate the code, but it would report an error during OpAccessChain validation.

> but I just wonder if it's possible to have any other nested type composed of CoopMatrix where this line would lead to a crash

When we have matrix in a structure in a structure SROA will manage to replace the outer structure with the internal one (yet without promoting cooperative matrix). With O0 it of cause won't happen, but we still won't crash here as yet for the outer structure we would have an alloca. MatrixTy in this case will be nullptr, so finishSROACooperativeMatrix would do nothing and the translator will still again fail during OpAccessChain validation. I'm not keen to handle this non-practical case for a W/A, instead I'd rather fix LLVM SROA, but it requires more time.